docs/spec/concepts.rst

```diff
@@ -3,97 +3,186 @@
 Type system concepts
 ====================
 
-.. _`union-types`:
+Static, dynamic, and gradual typing
+-----------------------------------
+
+A **statically-typed** programming language runs a type-checker before running
+a program. The program is required to be well-typed according to the language's
+type system. The type system assigns a type to all expressions in the language
+and verifies that their uses obey the typing rules. Normally, a program that is
+not well-typed (i.e., one that contains a type error) will not run. Java and
+C++ are examples of statically-typed object-oriented languages.
+
+A **dynamically-typed** programming language does not run a type-checker before
+running a program. Instead, it will check the types of values before performing
+operations on them at runtime. This is not to say that the language is
+"untyped". Values at runtime have a type and their uses obey typing rules. Not
+every operation will be checked, but certain primitive operations in the
+language such as attribute access or arithmetic will be. Python was
+historically a dynamically-typed language.
+
+**Gradual typing** is the name for a specific way to combine static and dynamic
+typing. It allows mixing static and dynamic checking at a fine level of
+granularity. For instance, in type-annotated Python, individual variables,
+parameters, and returns can optionally be given a static type. Data structures
+such as objects can have a mix of static and dynamic-only checking. As an
```

Comment:
Are we discussing two different concepts here? The concept that a given annotation can contain the "Dyn" type, but also the granularity required by the type-checker for mixing typed/untyped code?

Reply:
I'm not sure I understand the distinction between what "a given annotation can contain" vs "the granularity required by the type checker." Can you say more?

I see this section as a fairly informal introduction to the concept that Any and static types can coexist at fine granularity.

Reply:
I was more wondering what we are talking about from L26 to L32. Are we describing partial type annotations such as `Dyn -> Int` or are we still talking about granularity? also is there a distinction between variables and parameters vs. objects that we are trying to highlight? overall, I am just finding those lines hard to grasp